src/initialization/runpyC.C

```diff
@@ -0,0 +1,154 @@
+// export PYTHONPATH=.:$PYTHONPATH
+
+#define FORT(x) x##_
+
+static PyObject *py_module = 0;
+static PyObject *py_function = 0;
+
+inline int init_numpy()
+{
+  if(!PyArray_API) import_array();
+  return PyArray_API ? 1 : 0;
+}
+
+static void runpyC_1_1d(const char *py_script_, const char *py_function_,
+		             double *x, const int &n1)
+{
+  if(!Py_IsInitialized()) Py_Initialize();
+
+  assert(Py_IsInitialized());
+  assert(init_numpy());
+  printf("runpyC_1_1d: n1,%d \n", n1);  
+  printf("From runpyC_1_1d: Trying to call %s in %s \n",py_function_,py_script_);
+  if(!py_module) {
+    PyObject *py_script = PyUnicode_DecodeFSDefault(py_script_);
+    assert(py_script);
+    py_module = PyImport_Import(py_script);
+//Interesting: The following assertion fails if 
+//             the py_script has a python bug even outside the function that is called!
+//             the py_script has import torch
+    assert(py_module);
+    Py_DECREF(py_script);
+  }
+  if(!py_function) {
+    assert(py_module);
+    py_function = PyObject_GetAttrString(py_module, py_function_);
+    assert(py_function);// This assertation fails if py_function is not found in py_script
+    assert(PyCallable_Check(py_function));
+  }
+  //Simle tests
+  //assert(PyObject_CallFunctionObjArgs(py_function, NULL)); //This worked. Printed from the function.
+  
+  const npy_intp dim_x [] = { n1 };
+  PyObject *x_py = PyArray_SimpleNewFromData(1, dim_x, NPY_DOUBLE, x);
+  PyObject *pValue;
+  assert(pValue = PyObject_CallFunctionObjArgs(py_function, x_py, NULL));
+  if (pValue != NULL) {
+       printf("Result of call: %f\n", PyFloat_AsDouble(pValue));
+       Py_DECREF(pValue);
+  }
+  Py_DECREF(x_py); x_py = 0;
+  
+  // PyRun_SimpleString("import sys; sys.stdout.flush()");
+  
+  std::cout.flush();
+}
+
+static void runpyC_1_3d(const char *py_script_, const char *py_function_,
+			const char *varname_,
+			double *x, const int &n1, const int &n2, const int &n3 )
+{
+  if(!Py_IsInitialized()) Py_Initialize();
+
+  assert(Py_IsInitialized());
+  assert(init_numpy());
+  printf("runpyC_1_3d: varname %s , dims %d,%d,%d \n", varname_, n1,n2,n3);  
+  printf("From runpyC_1_3d: Trying to call %s in %s.py \n",py_function_,py_script_);
+  if(!py_module) {
+    PyObject *py_script = PyUnicode_DecodeFSDefault(py_script_);
+    assert(py_script);
+    py_module = PyImport_Import(py_script);
+//Interesting: The following assertion fails if 
+//             the py_script has a python bug even outside the function that is called!
+//             the py_script has import torch
+    assert(py_module && " This assertation fails if the py_script has a python bug..");
+    Py_DECREF(py_script);
+  }
+  if(!py_function) {
+    assert(py_module);
+    py_function = PyObject_GetAttrString(py_module, py_function_);
+    assert(py_function && " This assertation fails if the function is not found in the script."); //This assertation fails if py_function is not found in py_script
```

Comment:
It could also fail if the python script does not exist or is not found since PYTHONPATH is not set. I had to do this in the workdir   setenv PYTHONPATH `pwd`

Reply:
It could fail if for any reason the python script could not import a module. I had this assert fail because of "import matplotlib.pyplot as plt"  without giving any clue. It was working a few months ago when I tested it. Python script called on its own can import plt, but refuses when called through the exectable. numpy and torch can be imported without a problem.